[deepspeed] performance docs (#21573)
huggingface/transformers: docs/source/en/main_classes/deepspeed.mdx

   the increased data buffers.

#### ZeRO-0 Config

> I'm listing Stage 0 and 1 last since they are rarely used.

Reviewer: I'm not sure what this brings to the doc. Let's remove this sentence?

Author: Perfect, thank you, Sylvain.

}
```

#### How to Choose Which ZeRO Stage and Offloads To Use For Best Performance

> So now you know there are all these different stages and how do I know which of them to use. This section will attempt to address this question.

Reviewer: ```suggestion
So now you know there are all these different stages. How to decide which of them to use? This section will attempt to address this question.
```

2. Try ZeRO stage 2 first. if OOM then
3. Try ZeRO stage 2 + `offload_optimizer` - if OOM then
4. Switch to ZeRO stage 3 - if OOM then
5. Enable `offload_param` to `cpu` - if OOM then
6. Enable `offload_optimizer` to `cpu` - if OOM then

7. If you still can't fit a BS=1 first check various default values and lower them if you can. For example, if you use `generate` and you don't use a wide search beam make it narrower as it'd take a lot of memory.

8. Definitely use mixed half-precision over fp32 - so bf16 on Ampere and higher GPUs and fp16 on older gpu architectures.

9. If you still OOM you could add more hardware or enable ZeRO-Infinity - that is switch offloads `offload_param` and  `offload_optimizer` to `nvme`. You need to make sure it's a very fast nvme. As an anecdote I was able to infer BLOOM-176B on a tiny GPU using ZeRO-Infinity except it was extremely slow. But it worked!

You can, of course, work through these steps in reverse by starting with the most GPU memory efficient config and then going backwards. Or try bi-secting it.

Once you have your BS=1 not leading to OOM, measure your effective throughput.

> Next try to increase the BS to as large as you can, since the higher the batch size the more efficient the GPUs are as they perform the best when matrices they multiply are huge.

Reviewer: We're in the doc, let's spell batch size since there is no character limit :-)

Now the performance optimization game starts. You can turn off some offload features or step down in ZeRO stages and increase/decrease BS and again measure your effective throughput. Rinse and repeat until satisfied.

Don't spend forever on it, but if you're about to start a 3 months training - do spend a few days on it to find the most effective throughput-wise setup. So that your training cost will be the lowest and you will finish training faster. In the current crazy-paced ML world, if it takes you an extra month to train something you are likely to miss a golden opportunity. Of course, this is only me sharing an observation and in no way I'm trying to rush you. Before beginning to train BLOOM-176B I spent 2 days on this process and was able to increase throughput from 90 to 150 TFLOPs! This effort saved us more than one month of training time.

> These notes were written primarily for the training mode, but they should mostly apply for inference as well. For example, during inference Gradient Checkpointing is a no-op since it is only useful during training.

Reviewer: If not training or not using DeepSpeed inference, we should recommend Accelerate, as it was faster in the tests.

Author: does this work?

```suggestion
These notes were written primarily for the training mode, but they should mostly apply for inference as well. For example, during inference Gradient Checkpointing is a no-op since it is only useful during training. Additionally, we found out that if you 
are doing a multi-GPU inference and not using [DeepSpeed-Inference](https://www.deepspeed.ai/tutorials/inference-tutorial/), [Accelerate](https://huggingface.co/blog/bloom-inference-pytorch-scripts) should provide a superior performance.

```
